libkbfs/random_backoff.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+)
+
+var randomBackoffRand *rand.Rand
+
+func init() {
+	randomBackoffRand = rand.New(rand.NewSource(time.Now().Unix()))
+}
+
+type fireOnce struct {
+	ch   chan struct{}
+	once sync.Once
+}
+
+func newFireOnce() *fireOnce {
+	return &fireOnce{ch: make(chan struct{})}
+}
+
+func (o *fireOnce) fire() {
+	if o == nil {
+		return
+	}
+	o.once.Do(func() { close(o.ch) })
+}
+
+func (o *fireOnce) wait() {
+	if o == nil {
+		return
+	}
+	<-o.ch
+}
+
+// CancelableRandomBackoff can be used to wait on a random backoff timer. A
+// pointer to a zero value of CancelableRandomBackoff if usable.
+type CancelableRandomBackoff struct {
+	mu sync.Mutex
+	ch *fireOnce
```

Comment:
Sorry for the confusion. `Timer.Stop` stops the timer, but doesn't close the channel. So it prevents it from firing at all. But at the same time, `CancelableRandomBackOff.Wait` is still waiting on `f.wait()`. We'd need a `Timer.Fire` method in order to use it.